src/lua/httpc.lua

```diff
@@ -320,6 +326,122 @@ local function encode_url_params(params)
+        if data ~= nil then
+            local len = ffi.cast('char *', data) - rbuf.rpos + #eol
+            if shortest == nil or shortest > len then
+                shortest = len
+            end
+        end
+    end
+    if shortest ~= nil and shortest <= limit then
+        return shortest
+    elseif limit <= rbuf:size() then
+        return limit
+    end
+    return nil
+end
+
+local function io_read(self, opts, timeout)
+    timeout = timeout or IO_TIMEOUT_DEFAULT
+    local chunk
+    local delimiter
+    local check
+    if type(opts) == 'number' then
+        chunk = opts
+        check = check_limit
+    elseif type(opts) == 'string' then
+        chunk = IO_CHUNK_DEFAULT
+        delimiter = {opts}
+        check = check_delimiter
+    elseif type(opts) == 'table' then
+        chunk = opts.chunk or IO_CHUNK_DEFAULT
+        delimiter = opts.delimiter
+        if delimiter == nil then
+            check = check_limit
+        elseif type(delimiter) == 'string' then
+            delimiter = {delimiter}
+            check = check_delimiter
+        elseif type(delimiter) == 'table' then
+            check = check_delimiter
+        end
+    else
+        error('Usage: io:read(delimiter|chunk|{delimiter = x, chunk = x}, timeout)')
+    end
+
+    if chunk < 0 then
+        error('io:read(): chunk can not be negative')
+    end
+
+    chunk = math.min(chunk, IO_CHUNK_DEFAULT)
+    local rbuf = self._internal.rbuf
+    if rbuf == nil then
+        rbuf = buffer.ibuf()
+        self._internal.rbuf = rbuf
+    end
+
+    local len = check(self, chunk, delimiter)
+    if len ~= nil then
+        local data = ffi.string(rbuf.rpos, len)
+        rbuf.rpos = rbuf.rpos + len
+        return data
+    end
+
+    local deadline = fiber.clock() + timeout
+    repeat
+        local to_read = math.min(chunk - rbuf:size(), buffer.READAHEAD)
+        local data = rbuf:reserve(to_read)
+
+        local res = self._internal.io:read(data, rbuf:unused(), timeout)
+        if res == 0 then -- eof
+            self._errno = nil
+            local len = rbuf:size()
+            local data = ffi.string(rbuf.rpos, len)
+            rbuf.rpos = rbuf.rpos + len
+            return data
+        else
+            rbuf.wpos = rbuf.wpos + res
+            local len = check(self, chunk, delimiter)
+            if len ~= nil then
+                self._errno = nil
+                local data = ffi.string(rbuf.rpos, len)
+                rbuf.rpos = rbuf.rpos + len
+                return data
+            end
+        end
+        timeout = deadline - fiber.clock()
+    until timeout < 0
+    return nil
+end
+
+local function io_write(self, data, timeout)
+    timeout = timeout or IO_TIMEOUT_DEFAULT
+    return self._internal.io:write(data, #data, timeout)
+end
+
+local function io_close(self, timeout)
+    timeout = timeout or IO_TIMEOUT_DEFAULT
+    self["status"], self["reason"] = self._internal.io:close(timeout)
+end
```

Comment:
From a user perspective it looks unusual for me to see a timeout in a close method. I briefly looked around the code and, it seems, it waits for all in-fly write requests from other fibers. It is correct? I would like to see some clarification in the documentation (doc request?) about this timeout: what exactly it waits? I would also want to know, what is a state of the `io` if `:close()` fail on a timeout. Is it closed or open?

A side question: whether `:close()` is idempotent? IOW, whether `:close()` on a closed object is no-op and successful?

I think that it is good property for :close()` to never fail, because it is unclear what to do with a failed close. (`:close()` may report an error for diagnostic purposes, but still release resources.)

Reply:
> From a user perspective it looks unusual for me to see a timeout in a close method. I briefly looked around the code and, it seems, it waits for all in-fly write requests from other fibers. It is correct?

In this case, we are waiting for a server to close the connection after we have already done:
https://curl.se/libcurl/c/CURLOPT_READFUNCTION.html

If you stop the current transfer by returning 0 "pre-maturely" (i.e before the server expected it, **like when you have said you will upload N bytes and you upload less than N bytes), you may experience that the server "hangs" waiting for the rest of the data that will not come.** 

A similar case may be with `Transfer-Encoding: chunked` (we think that we already done, but a server don't).

Those. actually it can be used like `it waits for all in-fly write requests from other fibers`, but in fact the main idea to force close the connection from the client-side.

Reply:
> A side question: whether :close() is idempotent? IOW, whether :close() on a closed object is no-op and successful?

Yes, since the connection is closed, we have nothing to wait for. So we just finish the call immediately.

> I think that it is good property for :close() to never fail, because it is unclear what to do with a failed close. (:close()` may report an error for diagnostic purposes, but still release resources.)

Now (after the update) the error is returned only if timeout < 0. It looks clear, but I can do to remove the error:
```C
timeout = timeout >= 0 ? timeout : 0;
```
But it seems to me better with an error in this case.

Reply:
>> I think that it is good property for :close() to never fail <...>
>
> Now (I make a little update) the error is returned only if timeout < 0. <...>

Of course, if incorrect arguments are passed, the method should report an error. I meant all other situations except incorrect arguments. I think that if we're going to make `:close()` always-successful in given sense we should also handle fiber cancelling.

BTW, `:close()` should fail all subsequent `:write()` calls. Oleg said that there is a window after `:close()`, when a next `:write()` may succeed. I would look on it and estimate the time frame (one event loop iterator? a network trip?). I would also take care to effort that is needed to resolve it. Can't say now, whether it worth it.

Regarding waiting in `:close()` for the server to close its side: got it, need some time to think around.

Reply:
We agreed to rename `:close()` to `:finish()` due to its semantic with waiting for an answer. See below.